### PR TITLE
feat: Implement `InProgressBuildsSyncStage`

### LINKED ISSUE #1293: CI Integrations: Implement `InProgressBuildsSyncStage`
**Description**
 Implement `InProgressBuildsSyncStage`. Update the `SyncStagesFactory` to create the `InProgressBuildsSyncStage`.

**Acceptance Criteria**
- [ ] `InProgressBuildsSyncStage` implemented.
- [ ] `SyncStagesFactory` creates the `InProgressBuildsSyncStage`.
- [ ] Test new functionality.

**Additional context**
[Builds synchronization document](https://github.com/platform-platform/monorepo/blob/master/metrics/ci_integrations/docs/06_builds_synchronization.md)

### PR DESCRIPTION
**Pull request by solid-daniyilyevtyushkin**
Fixes #1293

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 1]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+
+        verifyNever(destinationClient.updateBuilds(
+          destinationProjectId,
+          argThat(containsInProgressBuildMatcher),
+        ));
+      },
+    );
+
+    test(
+      ".call() does not update a build if it is null after refresh and does not exceed a timeout",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:235]:
```suggestion
      ".call() does not update a build if the corresponding build is null and the current one is not timed out",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 2]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+          predicate<BuildData>((build) {
+            return build is BuildData &&
+                build.buildNumber == inProgressBuild.buildNumber;
+          }),
+        );
+
+        verifyNever(destinationClient.updateBuilds(
+          destinationProjectId,
+          argThat(containsInProgressBuildMatcher),
+        ));
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:263]: Same as for the previous test case.

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 3]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+      },
+    );
+
+    test(
+      ".call() does not update a build if it does not exceed a timeout and refreshing this build fails",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:268]:
```suggestion
      ".call() does not update a build if fetching the corresponding build fails and the build is not timed out",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 4]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+          predicate<BuildData>((build) {
+            return build is BuildData &&
+                build.buildNumber == inProgressBuild.buildNumber;
+          }),
+        );
+
+        verifyNever(destinationClient.updateBuilds(
+          destinationProjectId,
+          argThat(containsInProgressBuildMatcher),
+        ));
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:296]: Same as for the previous test case.

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 5]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+      },
+    );
+
+    test(
+      ".call() times out a build if it is in-progress after refresh and exceeds a timeout",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:301]:
```suggestion
      ".call() updates a build as timed out if the corresponding build is in progress and the build exceeds a timeout duration",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 6]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+
+        verify(destinationClient.updateBuilds(
+          destinationProjectId,
+          [expectedTimedOutBuild],
+        )).called(once);
+      },
+    );
+
+    test(
+      ".call() times out a build if it is null after refresh and exceeds a timeout",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:331]:
```suggestion
      ".call() updates a build as timed out if the corresponding build is null and the build exceeds a timeout duration",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 7]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+
+        verify(destinationClient.updateBuilds(
+          destinationProjectId,
+          [expectedTimedOutBuild],
+        )).called(once);
+      },
+    );
+
+    test(
+      ".call() times out a build if it exceeds a timeout and refreshing this build fails",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:361]:
```suggestion
      ".call() updates a build as timed out if it exceeds a timeout duration and fetching the corresponding build fails",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 8]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+
+        verify(destinationClient.updateBuilds(
+          destinationProjectId,
+          [expectedTimedOutBuild],
+        )).called(once);
+      },
+    );
+
+    test(
+      ".call() updates a build if it is finished after refresh",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:391]:
```suggestion
      ".call() updates a build if the corresponding build is finished",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 9]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+        );
+
+        await syncStage.call(syncConfig);
+
+        verifyNever(sourceClient.fetchCoverage(argThat(isIn(updatedBuilds))));
+      },
+    );
+
+    test(
+      ".call() does not continue the sync if fetching coverage for a build fails with an error",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:516]:
```suggestion
      ".call() does not continue syncing if fetching coverage for a build fails with an error",
```

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 10]
```diff
@@ -52,8 +110,535 @@ void main() {
...
+          buildNumber: 1,
+          buildStatus: BuildStatus.inProgress,
+          startedAt: startedAtExceedingTimeout,
+        );
+        final anotherInProgressBuild = BuildData(
+          buildNumber: 2,
+          buildStatus: BuildStatus.inProgress,
+          startedAt: startedAtExceedingTimeout,
+        );
+        final inProgressBuilds = [inProgressBuild, anotherInProgressBuild];
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:586]: There is the `inProgressBuilds` variable for the group - should we use it somehow in the last test cases?

**Review by solid-vladlaba — CHANGES_REQUESTED** [review]
Well done, @solid-daniyilyevtyushkin @solid-yevheniiashchypeleva! I've added a couple of suggestions below - please let me know what you think!

File: metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart [hunk 11]
```diff
@@ -52,8 +109,491 @@ void main() {
...
+        );
+        whenFetchBuild(inProgressBuilds[0]).thenAnswer(
+          (_) => Future.value(updatedBuilds[0]),
+        );
+        whenFetchBuild(inProgressBuilds[1]).thenAnswer(
+          (_) => Future.value(updatedBuilds[1]),
+        );
+        when(
+          destinationClient.updateBuilds(projectId, updatedBuilds),
+        ).thenAnswer((_) => Future.value(null));
```
Comment (solid-vovabeloded) [inline metrics/ci_integrations/test/integration/ci/sync_stage/builds/in_progress_builds_sync_stage_test.dart:595]:
```suggestion
        ).thenAnswer((_) => Future.value());
```
Please, apply this suggestion to all similar situations.

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Great job, @solid-daniyilyevtyushkin & @solid-yevheniiashchypeleva! I've added a few suggestions here - please review them.

**Review by solid-vovabeloded — APPROVED** [review]
LGTM, thanks @solid-daniyilyevtyushkin & @solid-yevheniiashchypeleva!

**Review by solid-vladlaba — APPROVED** [review]
LGTM, thanks @solid-daniyilyevtyushkin @solid-yevheniiashchypeleva !:rocket:

**Pull request merged**